awx/main/dispatch/worker/callback.py

```diff
@@ -48,24 +48,32 @@ def __init__(self):
         self.buff = {}
         self.pid = os.getpid()
         self.redis = redis.Redis.from_url(settings.BROKER_URL)
+        self.redis.client_setname("callbackbrokerworker")
+        self.subsystem_metrics = s_metrics.Metrics(conn=self.redis)
         self.prof = AWXProfiler("CallbackBrokerWorker")
         for key in self.redis.keys('awx_callback_receiver_statistics_*'):
             self.redis.delete(key)
 
     def read(self, queue):
         try:
-            res = self.redis.blpop(settings.CALLBACK_QUEUE, timeout=1)
-            if res is None:
-                return {'event': 'FLUSH'}
-            self.total += 1
-            return json.loads(res[1])
+            with self.subsystem_metrics.conn.pipeline() as pipe:
+                res = self.redis.blpop(settings.CALLBACK_QUEUE, timeout=1)
```

Comment:
This *could* get expensive doing this per-read.  That said, I can't think of an alternative here that won't result in potentially inaccurate data.

I'd be curious to see what happens performance-wise here if we attempt to throw really high event load at this PR.

Reply:
@chrismeyersfsu @fosterseth

Here's how I generally test this performance-wise for bottlenecks that can result in regressions.

First, I spun up a `devel` install, and stopped the callback receiver:

```shell
~ supervisorctl stop tower-processes:awx-receiver
```

Next, I ran a Python script in `~ awx-manage shell_plus` to fill up the redis callback queue (this takes a minute or two):

```python
In [1]: from awx.main.queue import CallbackQueueDispatcher

In [2]: import uuid

In [3]: d = CallbackQueueDispatcher()

In [4]: for i in range(1000000):
   ...:     d.dispatch({ 'uuid': str(uuid.uuid4()), 'stdout': 'Line %d' % i, 'job_id': 1, })
   ...:
```

```shell
~ docker exec -it tools_redis_1 bash
root@0a52a640e1d2:/data# redis-cli -s /var/run/redis/redis.sock
redis /var/run/redis/redis.sock> llen callback_tasks
(integer) 1000000
```

Finally, I ran `~ awx-manage callback_stats` and restarted the callback receiver w/ `~ supervisorctl stop tower-processes:awx-receiver`, and I watched to see how high my throughput got.  On my local Docker-based install I got pretty far:

```shell
bash-4.4$ awx-manage callback_stats
main_jobevent
↳  last minute 156992
```

(so about 8 to 9M inserts per hour)

Reply:
Next, I stopped all my containers, and re-ran this entire exercise using _this_ branch.  Unfortunately, these constant `incr()` calls really seem to drastically affect write throughput:

```shell
bash-4.4$ awx-manage callback_stats
main_jobevent
↳  last minute 76796
```

Reply:
We would probably make up a lot of this performance loss by *not* doing the `incr()` per `.read()`, but instead at the end of each batch.